R/tm_g_distribution.R

```diff
@@ -1282,37 +1290,40 @@ srv_distribution <- function(id,
 
     ### REPORTER
     if (with_reporter) {
-      card_fun <- function(comment, label) {
-        card <- teal::report_card_template(
-          title = "Distribution Plot",
-          label = label,
-          with_filter = with_filter,
-          filter_panel_api = filter_panel_api
-        )
-        card$append_text("Plot", "header3")
-        if (input$tabs == "Histogram") {
-          card$append_plot(dist_r(), dim = pws1$dim())
-        } else if (input$tabs == "QQplot") {
-          card$append_plot(qq_r(), dim = pws2$dim())
-        }
-        card$append_text("Statistics table", "header3")
-
-        card$append_table(common_q()[["summary_table"]])
-        tests_error <- tryCatch(expr = tests_r(), error = function(e) "error")
-        if (inherits(tests_error, "data.frame")) {
-          card$append_text("Tests table", "header3")
-          card$append_table(tests_r())
-        }
-
-        if (!comment == "") {
-          card$append_text("Comment", "header3")
-          card$append_text(comment)
-        }
-        card$append_src(teal.code::get_code(output_q()))
-        card
-      }
-      teal.reporter::simple_reporter_srv("simple_reporter", reporter = reporter, card_fun = card_fun)
+      card_function <- hydrate_function(card_function, with_filter, filter_panel_api)
+      teal.reporter::simple_reporter_srv("simple_reporter", reporter = reporter, card_fun = card_function)
     }
     ###
   })
 }
+
+#' @keywords internal
+tm_g_distribution_card_function <- function(comment, label) { #nolint: object_length.
```

Comment:
my main concern is that this is not a true list of required arguments. We are using `with_filter` and `filter_panel_api`. Moreover, we are using `dist_r()`, `pws1` and many more object from the parent environment.
You have addressed this with `hydrate_function` but I think it would be better not to be in need of such functionality. I can think of a few ways how to do this:
- extend function arguments
- pass data as environment
- this being a module?

(I really hope that there is no strict check for these and only these argument somewhere in the teal.reporter).

Each has its pros and cons and we probably need to think more which one would be best for this task. Glad you stopped early to allow for a discussion like this.

Looking at the changes - this is how it was written in the past so I definitely not blaming you for this. This PR is a great opportunity to make it right.

Reply:
> (I really hope that there is no strict check for these and only these argument somewhere in the teal.reporter).

[There is one, actually](https://github.com/insightsengineering/teal.reporter/blob/ad7fd2041f0d7cac83d0d97dbbe3ed7914a197ea/R/AddCardModule.R#L89). I considered just adding `...` to the card function's formals but that is forbidden. I think it would work but I am not sure about resolving conflicts between the variables passed to the ellipsis and bindings in the card function's parent environment. If we can figure out a way for the card functions to be pure, this would not be a problem.

I tried to limit the proposed changes to the module packages because I assumed modifying `teal.reporter` is off the table.

Note that passing the caller environment is not sufficient, at least in this module (I assume in others as well). `with_filter` and `filter_panel_api` are also required but they exist in the caller's parent frame, not in the caller itself.

I will be happy to discuss a satisfactory solution.

----

Note also that with the proposed solution the `if (with_reporter)` chunk would be the same in every module, which means _all code added to the module to enable reporting _would be the same in every module. The only difference would be the default card function, which is separate.

This in turn opens another possibility: have a function called `with_reporter` that would modify any module object by adding the code required for reporting into the module functions. Then adding support for reporting would be limited to something like `modules(tm_g_distribution(...) |> with_reporter(card_fun = foo))`. But that's a separate conversation.

Reply:
Ahh that's bad :( Then I would say that please feel free to modify this in `teal.reporter`. I feel that this is the true root-cause of the issue. I also think that this (relaxing this check on formals) would make creating card-factory-fun much simpler and also encapsulated so that it can be run without any additional steps - just provide what's required.
That's exactly what I'm looking for - more simplicity. I'm a little bit afraid of introducing a requirement of _hydrating_ a card-factory-fun. I don't think that most of our users (in particular: not developers) would handle this.

Reply:
Have a look at this alternative: https://github.com/insightsengineering/teal.modules.general/pull/742

I don't think it is full proof because of the way `teal.reporter` checks and handles formal arguments of `card_fun`:
https://github.com/insightsengineering/teal.reporter/blob/ad7fd2041f0d7cac83d0d97dbbe3ed7914a197ea/R/AddCardModule.R#L166

Note that environment shenanigans are already built into `teal.reporter`:
https://github.com/insightsengineering/teal.reporter/blob/ad7fd2041f0d7cac83d0d97dbbe3ed7914a197ea/R/AddCardModule.R#L173

I think this branch is the superior solution (not necessary the ideal one).

Reply:
Haven't yet checked the alternative, but in this case I would opt for relaxing teal.reporter checks to allow passing ellipsis which would simply the process and would not require the usage of hydrating in this case.

Reply:
passing `env` instead of `...` also works!